Lamina/include/Math/vector.hpp

```diff
@@ -0,0 +1,450 @@
+			}
+
+			this->data = other.data;
+			return *this;
+		}
+
+		std::array<t_vector, t_vec_size> data;
+	};
+
+	//specialization for 2d vector
+	template<typename t_vector>
+	class Vector<t_vector, 2>
+	{
+	public:
+		Vector() : data{} {};
+		Vector(const std::array<t_vector, 2>& _vector) { CreateVector(_vector); }
+		void CreateVector(const std::array<t_vector, 2>& _vector)
+		{
+			if (_vector.size() == data.size()) std::copy_n(_vector.begin(), 2, data.begin());
+			else std::cout << "Unable to assign vector. Incompatible size" << std::endl;
+		}
+
+		Vector operator+ (const t_vector other)
+		{
+			lm::Vector<t_vector, 2> result(this->data);
+			for (int i = 0; i < this->data.size(); i++) result.data[i] += other;
+			return result;
+		}
+		Vector operator+ (Vector& other)
+		{
+			if (other.data.size() == 2)
+			{
+				lm::Vector<t_vector, 2> result(this->data);
+				for (int i = 0; i < this->data.size(); i++) result.data[i] += other.data[i];
+				return result;
+			}
+			else std::cout << "Incompatible vector size for addition\n";
+		}
+		Vector operator+= (Vector& other)
+		{
+			if (this->data.size() != other.data.size())
+			{
+				std::cout << "Incompatible vector size for addition\n";
+				return *this;
+			}
+			for (int i = 0; i < this->data.size(); i++)
+			{
+				this->data[i] += other.data[i];
+			}
+			return *this;
+		}
+		Vector operator- (const t_vector other)
+		{
+			for (int i = 0; i < data.size(); i++) data[i] -= other;
+			return *this;
+		}
+		Vector operator-= (Vector& other)
+		{
+			if (other.data.size() == 3)
+			{
+				for (int i = 0; i < this->data.size(); i++) this->data[i] -= other.data[i];
+				return *this;
+			}
+			else std::cout << "Incompatible vector size for addition\n";
+		}
+		Vector operator- (const Vector& other)
+		{
+			if (other.data.size() == 3)
+			{
+				lm::Vector<t_vector, 3> result(this->data);
+				for (int i = 0; i < this->data.size(); i++) result.data[i] -= other.data[i];
+				return result;
+			}
+			else std::cout << "Incompatible vector size for addition\n";
+		}
+
+		Vector operator* (const t_vector& scalar) {
+			return lm::Vector<t_vector, 2>({
+				this->data[0] * scalar,
+				this->data[1] * scalar,
+				});
+		}
+
+		Vector operator = (const Vector& other)
+		{
+			if (this == &other)
+			{
+				return *this;
+			}
+
+			this->data = other.data;
+			return *this;
+		}
+
+		std::array<t_vector, 2> data;
+
+		inline t_vector& x() { return data[0]; }
+		inline t_vector& y() { return data[1]; }
+	};
+
+	//specialization for 3d vector
+	template<typename t_vector>
+	class Vector<t_vector, 3>
+	{
+	public:
+		Vector() : data{} {};
+		Vector(const std::array<t_vector, 3>& _vector) { CreateVector(_vector); }
+		void CreateVector(const std::array<t_vector, 3>& _vector)
+		{
+			if (_vector.size() == data.size()) std::copy_n(_vector.begin(), 3, data.begin());
+			else std::cout << "Unable to assign vector. Incompatible size" << std::endl;
+		}
+
+		Vector operator+ (const t_vector other)
```

Comment:
Vector should be passed by reference (same issue is present in a few functions below).